proposals/interface-types/working-notes/scenarios/chooser.md

```diff
@@ -0,0 +1,680 @@
+  i32.const 1                ;; initial reference count of 1
+  call-export "shared-builder-x" ;; make it a shared ptr pair
+  own (i32)                  ;; re-own it, with a different destructor
+    call-export "shared-release-x"
+  end
+end
+)
+
+(@interface func (export "chooser")
+  (param $left string)
+  (param $right string)
+  (result string)
+
+  local.get $left
+  invoke-func $stralloc-x
+  invoke-func $shared-x     ;; make it a shared ptr
+  let (local $lpo owned<i32>) (result string)
+    local.get $right
+    invoke-func $stralloc-x
+	invoke-func $shared-x   ;; make it a shared ptr
+    let (local $rp owned<i32>) (result string)
+      ;; set up the call to the core chooser 
+      local.get $lp
+	  owned.access
+      local.get $rp
+	  owned.access
+      call-export "nondeterministic_choice"  ;; call the chooser itself
+      own (i32)               ;; own the result
+        call-export "shared-release-x"    ;; will eventually call to free string
+      end		
+      owned.access
+	  call-export "shared-get-x"
+	  call-export "access-utf8-x"
+      string.from.utf8
+    end
+  end
+)
+```
+
+For clarity, we separate out two helper functions -- `shared-x` and `stralloc-x`
+-- whose role is to create a shared pointer pair and to instantiate a memory
+string from an Interface Types's `string` entity.
+
+>Note the `-x` suffix signals that these functions are part of the exporting
+>module.
+
+>Note, like the import and export adapters themselves, they will be inlined as
+>part of the adapter fusion process.
+
+The returned value from `nondeterministic_choice` is wrapped up as an `own`ed
+allocation and lifted to a `string`.
+
+>Note we do not need to `own` the string memory of the arguments because we have
+>asserted that both the arguments to `nondeterministic_choice` will be the
+>_last_ references to the strings. However, only one of the C++ strings will be
+>deallocated -- the other is returned to us. We _do_ need to `own` the return
+>result, however.
+
+In addition to the Interface Types memory management, because we are using a
+non-trivial C++ structure, we have to invoke the appropriate constructors,
+access functions and destructors of our `shared_ptr` structure.
+
+This is achieved through some additional core helper functions that are exported
+by the core wasm module, although they would not typically be exported as
+Interface Type functions:
+
+* `shared-builder-x` is used to create and initialize a `shared_ptr`
+  structure. It takes a pointer as an argument and returns a `shared_ptr` with
+  an initial reference count of 1.
+
+* `shared-release-x` is used to decrement the reference count of a `shared_ptr`;
+  and if that results in the reference count dropping to zero _both the
+  underlying resource and the `shared_ptr` structure itself are disposed of_.
+
+* `shared-get-x` is used to access the underlying resource of a `shared_ptr`
+  without affecting it's reference count.
```

Comment:
nit: its

Reply:
My excuse: English is my second language :)